### PR TITLE
add protocol to request retrieval ask

### PR DESCRIPTION
**Pull request by dirkmc**
Required by https://github.com/filecoin-project/lotus/issues/5820

### CODE REVIEW

**Review by nonsense — APPROVED** [review]

File: retrievalmarket/impl/provider.go [hunk 1]
```diff
@@ -275,6 +276,42 @@ func (p *Provider) ListDeals() map[retrievalmarket.ProviderDealIdentifier]retrie
 	return dealMap
 }
 
+// HandleAskStream is called when the provider receives a request for the
+// retrieval ask (how much the provider charges for retrieval)
+func (p *Provider) HandleAskStream(s rmnet.RetrievalAskStream) {
```
Comment (aarshkshah1992) [inline retrievalmarket/impl/provider.go:281]: @dirkmc  This wont work. With Dynamic pricing, we quote a different price for each deal based on a number of factors. Just returning the ask from the ask store wont work. Maybe the storage provider has an external script configured for example. You want the client to give you the params it gives you when it actually makes the deal (pieceCid etc.) in the request for this stream and then run it through the same pricing function that we run during retrieval.

File: retrievalmarket/types.go [hunk 2]
```diff
@@ -120,6 +124,12 @@ type RetrievalPeer struct {
 	PieceCID *cid.Cid
 }
 
+// SignedRetrievalAsk is an ask signed by the miner's private key
+type SignedRetrievalAsk struct {
```
Comment (aarshkshah1992) [inline retrievalmarket/types.go:128]: @dirkmc  We don't really sign the QueryResponse for the query protocol. So, why are we signing the ask response here ?

File: retrievalmarket/impl/askstore/askstore_impl.go [hunk 3]
```diff
@@ -79,7 +99,40 @@ func (s *AskStoreImpl) GetAsk() *retrievalmarket.Ask {
 		return nil
 	}
 	ask := *s.ask
-	return &ask
+	return ask.Ask
+}
+
+// GetSignedAsk returns the current retrieval ask, signed with the provider's
+// key, or nil if there is no ask set.
+func (s *AskStoreImpl) GetSignedAsk() (*retrievalmarket.SignedRetrievalAsk, error) {
```
Comment (aarshkshah1992) [inline retrievalmarket/impl/askstore/askstore_impl.go:107]: Why do we need to sign asks ? Also, as commented elsewhere, just returning the ask here without using the pricing function wont be of much use.

**Review by aarshkshah1992 — CHANGES_REQUESTED** [review]

File: retrievalmarket/impl/client.go [hunk 4]
```diff
@@ -178,6 +178,56 @@ func (c *Client) FindProviders(payloadCID cid.Cid) []retrievalmarket.RetrievalPe
...
+	}
+
+	// Open a stream to the provider over the ask protocol
+	s, err := c.network.NewAskStream(ctx, p.ID)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to open stream to miner: %w", err)
+	}
+
+	// Send a request for the provider's ask
+	request := rmnet.AskRequest{Miner: p.Address}
```
[the remark this answers is not in the document]
Comment (dirkmc) [reply retrievalmarket/impl/client.go:198]: We need to put _some_ parameter in the ask request in order to CBOR marshall it

[the remark this answers is not in the document]
Comment (dirkmc) [reply retrievalmarket/impl/client.go:198]: I'll change this to the parameters needed for dynamic retrieval pricing

**Comment by dirkmc** [comment]
Closing as this functionality doesn't add anything new over the `Query` protocol (see `HandleQueryStream`)

**Pull request closed without merging**